File: include/mjx/open.cpp
#include "mjx/open.h"

#include "mjx/internal/open.h"
#include "mjx/internal/types.h"

namespace mjx {

int Open::EventType(std::uint16_t bits) {
  auto open_type = internal::Open(bits).Type();
  if (open_type == internal::OpenType::kChi) {
    return static_cast<int>(mjxproto::EVENT_TYPE_CHI);
  }
  if (open_type == internal::OpenType::kPon) {
    return static_cast<int>(mjxproto::EVENT_TYPE_PON);
  }
  if (open_type == internal::OpenType::kKanClosed) {
    return static_cast<int>(mjxproto::EVENT_TYPE_CLOSED_KAN);
  }
  if (open_type == internal::OpenType::kKanOpened) {
    return static_cast<int>(mjxproto::EVENT_TYPE_OPEN_KAN);
  }
  if (open_type == internal::OpenType::kKanAdded) {
    return static_cast<int>(mjxproto::EVENT_TYPE_ADDED_KAN);
  }
}

int Open::From(std::uint16_t bits) {
  return static_cast<int>(internal::Open(bits).From());
};

int Open::At(std::uint16_t bits, std::size_t i) {
  return internal::Open(bits).At(i).Id();
}

std::size_t Open::Size(std::uint16_t bits) {
  return internal::Open(bits).Size();
}

std::vector<int> Open::Tiles(std::uint16_t bits) {
  std::vector<int> tiles;
  for (const internal::Tile& t : internal::Open(bits).Tiles()) {
    tiles.push_back(t.Id());
  }
  return tiles;
}

std::vector<int> Open::TilesFromHand(std::uint16_t bits) {
  std::vector<int> tiles;
  for (const internal::Tile& t : internal::Open(bits).TilesFromHand()) {
    tiles.push_back(t.Id());
  }
  return tiles;
};

int Open::StolenTile(std::uint16_t bits) {
  return internal::Open(bits).StolenTile().Id();
}

int Open::LastTile(std::uint16_t bits) {
  return internal::Open(bits).LastTile().Id();
}

std::vector<int> Open::UndiscardableTileTypes(std::uint16_t bits) {
  std::vector<int> tile_types;
  for (const internal::TileType& tt :
       internal::Open(bits).UndiscardableTileTypes()) {
    tile_types.push_back(static_cast<int>(tt));
  }
  return tile_types;
}

}  // namespace mjx

File: mjx/open.py
from __future__ import annotations

from typing import List

import _mjx  # type: ignore

from mjx.const import EventType, RelativePlayerIdx, TileType
from mjx.tile import Tile


class Open:
    def __init__(self, bit: int):
        self.bit = bit

    def event_type(self) -> EventType:
        return _mjx.Open.event_type(self.bit)

    def steal_from(self) -> RelativePlayerIdx:
        return RelativePlayerIdx(_mjx.Open.steal_from(self.bit))

    def at(self, i: int) -> Tile:
        return Tile(_mjx.Open.at(self.bit, i))

    def size(self) -> int:
        return _mjx.Open.size(self.bit)

    def tiles(self) -> List[Tile]:
        return [Tile(t) for t in _mjx.Open.tiles(self.bit)]

    def tiles_from_hand(self) -> List[Tile]:
        return [Tile(t) for t in _mjx.Open.tiles_from_hand(self.bit)]

    def stolen_tile(self) -> Tile:
        return Tile(_mjx.Open.stolen_tile(self.bit))

    def last_tile(self) -> Tile:
        return Tile(_mjx.Open.last_tile(self.bit))

    def undiscardable_tile_types(self) -> List[TileType]:
        return [TileType(t) for t in _mjx.Open.undiscardable_tile_types(self.bit)]
